homeassistant/helpers/script.py

```diff
@@ -390,47 +423,108 @@ def async_script_wait(entity_id, from_s, to_s):
         if not unsub:
             return
 
-        if not self._stop.is_set():
-            self._changed()
+        self._changed()
         try:
-            timeout = self._action[CONF_TIMEOUT].total_seconds()
+            delay = self._action[CONF_TIMEOUT].total_seconds()
         except KeyError:
-            timeout = None
+            delay = None
         done = asyncio.Event()
         try:
-            await asyncio.wait_for(
-                asyncio.wait(
+            async with timeout(delay):
+                _, pending = await asyncio.wait(
                     {self._stop.wait(), done.wait()},
                     return_when=asyncio.FIRST_COMPLETED,
-                ),
-                timeout,
-            )
+                )
+            for pending_task in pending:
+                pending_task.cancel()
         except asyncio.TimeoutError:
             if not self._action.get(CONF_CONTINUE, True):
                 self._log(_TIMEOUT_MSG)
                 raise _StopScript
         finally:
             unsub()
 
+    async def _async_call_service_step(self):
+        """Call the service specified in the action."""
+        domain, service, service_data = self._prep_call_service_step()
+        blocking = service_data.get(ATTR_BLOCKING, True)
```

Comment:
This prevents us from passing `blocking` to the service that we're going to call. It also means that we can end up passing a variable to a service that is not valid for the service schema. We need to find a different way of passing `blocking` here. 

I also think that specifying `blocking` here is maybe out of scope of this PR? It seems like an unrelated change from specifying the run mode ?

Reply:
Believe me, this is the area that caused me the most trouble and time. I tried at least three different approaches:

1. The original `run_mode` script helper parameter.
1. Modify the basic service call mechanism so that _any_ service could be called in the background.
1. Modify the script helper's service handling method to allow _any_ service to be called in the background.

All of these had practical issues (which are too difficult to explain in a short comment.)

I found this to be the most practical way to handle it. And it's in scope because this method has to use this information to determine what to set the call limit to. (The rest of the implementation is in the script integration, which I already have working.)

And this doesn't "pass" the parameter to any services. It only looks for it if it's there, and _then_ only uses it if the service is `script.turn_on`. (You'll notice it still sets `blocking=True` when it gets to the `hass.services.async_call` invocation.) I don't see how this can interfere with any other services. Again, it's not passing it to anything, it's only using it if it's there. (BTW, it's being added to the `script.turn_on` service's schema. Again, that's already in the script integration modifications I've already made and tested and will be in the next PR.)

Reply:
Ah I see it now. I didn't realize that it was only used inside the if-statement.

Why don't we just always call script domain without a limit ? If it's a non blocking run, it will just return very fast and we're ok?

Reply:
Yes, that would work. I just assumed the limit would be useful, but I suppose that's probably really more for faulty integrations as opposed to core things. So, yeah, I can just disable the limit for all `script` services, which simplifies this a bit and decouples it from the `script.turn_on` service implemented in the `script` integration. Nice.